### PR TITLE
perf: fused RMSNorm (5.96x), Snake torch.compile (1.61x), Euler pre-alloc (1.8x)

### PR DESCRIPTION
**Pull request by mstkyvz**
## Summary

Three kernel-level performance optimizations measured on H100 80GB (PyTorch 2.12, CUDA 13.0). All numbers are wall-clock means over 3000 iterations with `torch.cuda.synchronize()`, using the actual class-method call path that matches production usage.

---

### 1. `layers/layernorm.py` — Replace manual RMSNorm with `F.rms_norm`

`torch.nn.functional.rms_norm` dispatches to PyTorch's native fused CUDA kernel, eliminating the 4 separate kernel launches of the current `@torch.compile` implementation.

Two separate call paths are optimized differently:

**`rms_forward`** — pass input directly; fused kernel handles precision internally:
```python
return F.rms_norm(x, (self.hidden_size,), self.weight, self.eps)
```

**`add_rms_forward`** — compute residual sum in float32 (matching original precision), then normalize:
```python
orig_dtype = x.dtype
x = x.float() + residual.float()
residual = x.to(orig_dtype)
return F.rms_norm(x, (self.hidden_size,), self.weight.float(), self.eps).to(orig_dtype), residual
```

| Variant | Before | After | Speedup |
|---|---|---|---|
| `rms_forward` | 0.0641 ms | 0.0107 ms | **5.96x** |
| `add_rms_forward` | 0.0661 ms | 0.0485 ms | **1.36x** |

With 28 layers × 2 norms per layer × 2 LMs (base + residual) = **112 RMSNorm calls per autoregressive step**, this saves roughly **5.9 ms per token step** at batch=1.

**Numerical accuracy** (verified against float64 ground truth):

| Variant | Max diff vs original | ULP |
|---|---|---|
| `rms_forward` | 1.953e-03 | 2.0 ULP |
| `add_rms norm output` | 9.766e-04 | 1.0 ULP |
| `add_rms residual` | 0.000 | bit-exact |

Both implementations agree to within 1–2 ULP in float16 — this is float16 precision noise and has no effect on model output quality.

---

### 2. `layers/audio_vae_v2.py` — Replace `@torch.jit.script` with `@torch.compile` on `snake()`

`torch.compile` applies more aggressive operator fusion and buffer reuse than TorchScript.

| | Before (`jit.script`) | After (`compile`) | Speedup |
|---|---|---|---|
| `snake()` — 1×2048×12000 fp32 | 0.1109 ms | 0.0689 ms | **1.61x** |

Max numerical diff: 0.00e+00 (bit-exact on every tested seed).

---

### 3. `models/voxcpm2/model.py` — Pre-allocate CFM Euler buffers before the timestep loop

The current `solve_euler` calls `torch.zeros(...)` for five buffers inside each Euler step. Moving allocation outside the loop (`torch.empty` once, explicit `zero_()` only where the unconditioned branch requires it) reduces CUDA allocations from `5 × n_timesteps` to 5 per inference.

| | Before | After | Speedup |
|---|---|---|---|
| Allocation overhead (10 steps) | 0.3638 ms | 0.2073 ms | **1.76x** |

Correctness: bit-exact across all batch sizes (bsz=1,4,8) and all 10 timesteps.

---

## Compatibility

- `F.rms_norm` requires PyTorch ≥ 2.4 — consistent with the `torch.compile` usage already in the codebase.
- `@torch.compile` on a module-level function is a drop-in replacement for `@torch.jit.script`; signature and behaviour are identical.
- The Euler pre-alloc preserves all write patterns and initialisation semantics of the original.

### CODE REVIEW

**Review by a710128 — APPROVED** [review]
LGTM! Nicely scoped and well-benchmarked. 🚀

I verified the numerical claims locally (torch 2.10, CUDA) and everything checks out:

- **RMSNorm → `F.rms_norm`**: In the model's actual bf16 config, the new path is actually *more* accurate than the original (0.0099 vs 0.0234 error vs fp64 ground truth), since it keeps the weight multiply in fp32 internally. Residual is bit-exact. dtypes match, so the fused path is used. ✅
- **Euler pre-alloc**: The one risky bit (`mu_in`'s unconditioned half previously relying on `torch.zeros`) is correctly handled by the added `mu_in[bsz:].zero_()`. All other buffers are fully overwritten each step, so `torch.empty` is safe. ✅
- **Snake `jit.script` → `compile`**: Function body unchanged, bit-exact — just a better compile backend. ✅

A few non-blocking nits for whenever you get a chance:
1. The `add_rms_forward` speedup differs between the commit message (4.2x) and the description table (1.36x) — worth reconciling (the 1.36x table looks like the final number).
2. `snake(x, alpha)` lost its type hints in the switch away from `jit.script` — a `torch.Tensor` annotation would match the project convention.
3. Maybe a one-line comment in `layernorm.py` noting that weight/activation dtypes must match to keep the fused path (guards against a future silent perf regression).

Approving — thanks for the clean work! 🙌
